### DISCUSSION TITLE
Semantics of `ErrMode::Incomplete` and `Needed`: need some bytes for what goal?

### DISCUSSION THREAD

**Discussion by lo48576** [winnow-rs/winnow#208] [Q&A]
Streaming parser with incomplete input will need some more bytes, but what is the goal when the parser needs that length of bytes...  **to change the parser state**, or **to complete the parsing successfully**?

Consider you are writing a parser for the Rust-style character literal syntax, such as `'a'` and `'\u{0}'`.
If the parser get the incomplete stream with the content `'`, what the parser should return?

* To successfully complete the parsing, the parser needs at least 2 more bytes (for example, `a` and closing `'`).
* To get to the new parsing state, one more byte (at least) is enough: if the next character is `'`, the parser can fail since `''` is invalid.

And if the parser get `'\`, how many bytes the parser should return as `Needed`?

* To successfully complete the parsing, the parser needs at least 2 more bytes (for example, `n'`).
* To get to the new parsing state, one more byte (at least) is enough:
    + If the next character is `n`, the possible valid string is now limited to `'\n'`.
    + If the next character is `\`, the possible valid string is now limited to `'\\'`.
    + If the next character is `u`, the parser can call the new inner parser for Unicode codepoint (possibly `delimited('{', hex_digit1, '}')` parser).

Documentation for [`ErrMode`](https://docs.rs/winnow/0.3.6/winnow/error/enum.ErrMode.html) (as of v0.3.6) describes as below, but it is not clear enough for me.

> More data needs to be buffered before retrying the parse.

"Retrying the parse"... to complete parsing in success? Or to return any meaningful result including backtrack and cut?

So I'd like to know what should `Needed` value mean in the context of `ErrMode::Incomplete`.
My personal preference is, "returning minimum bytes needed to possibly return non-`Incomplete` result", because it will make the value useful for optimization since it possibly lets `alt` parser wait fewer tokens to try the next parser.

**Comment by lo48576** [comment]
Ah... "returning minimum bytes needed to possibly return non-`Incomplete` result" is not precisely what I intended, since another inner parser after the new state will return `Incomplete` independently.
I'm not sure how I should say what I think, but you know what I mean.

**Comment by epage** [comment]
> Streaming parser with incomplete input will need some more bytes, but what is the goal when the parser needs that length of bytes... to change the parser state, or to complete the parsing successfully?

Before I get into my thoughts, some context
- This design is inherited from `nom` which `winnow` is forked from
- I've not directly  needed streaming parsing but understand some of the use cases, so mine is more of an armchair analysis

I think the best answer is the most practical: if incomplete parsing just strings you along from parser to parser as you slowly progress parse states a byte or two at a time, that would be an inefficient program.  Most likely, you will be wanting to read larger chunks at a time.  For this reason, I find the only useful information to be either a generic "more is needed" or an exact amount **if** reported by a header (like with `length_data` and friends).  See also [the partial parsing special topic](https://docs.rs/winnow/latest/winnow/_topic/partial/index.html).

I have not felt confident enough in this to make any documentation updates yet, instead hoping I could collect input from users like in #169 to build up the appropriate context for what streaming parsing *should* look like.

**Comment by lo48576** [comment]
> if incomplete parsing just strings you along from parser to parser as you slowly progress parse states a byte or two at a time, that would be an inefficient program.

I agree, practical programs in usual situation won't retry parsing after putting just one or two bytes to the input buffer.

Unconditionally returning `ErrMode::Incomplete(Needed::Unknown)` won't cause troubles, but always using it everywhere doesn't feel correct to me... since it is named "unknown", not "unspecified" or something.

**Reply by epage** [reply]
Names can be changed to clarify :)

But I'm not too worried about that name because, in the larger picture, it is unknown.  The parser returning it only knows for itself what is needed but that is an incomplete picture.

For `nom`, someone did propose just switching to `Option` which has some appeal for access to combinators (rust-bakery/nom#1159)
